apis/Google.Cloud.Firestore/Google.Cloud.Firestore/Filter.cs

```diff
@@ -0,0 +1,240 @@
+
+    /// <summary>
+    /// 
+    /// </summary>
+    internal class UnaryFilter : Filter, IEquatable<UnaryFilter>
+    {
+        public FieldPath Field { get; }
+
+        // The integer value of either the UnaryOp or FieldOp, depending on whether _value is null.
+        public int Op { get; }
+
+        // The value for a field (binary) operator, or null for a unary operator
+        public Object Value { get; }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="fieldPath"></param>
+        /// <param name="op"></param>
+        /// <param name="value"></param>
+        public UnaryFilter(FieldPath fieldPath, FieldOp op, Object value)
+        {
+            Field = fieldPath;
+            var unaryOperator = GetUnaryOperator(value, op);
+            if (unaryOperator != UnaryOp.Unspecified)
+            {
+                Op = (int) unaryOperator;
+                Value = null;
+            }
+            else
+            {
+                Op = (int) op;
+                Value = value;
+            }
+        }
+
+        internal override ProtoFilter ToProto(SerializationContext context)
+        {
+            
+            return Value == null
+                ? new ProtoFilter { UnaryFilter = new StructuredQuery.Types.UnaryFilter { Field = Field.ToFieldReference(), Op = (UnaryOp) Op } }
+                : new ProtoFilter { FieldFilter = new StructuredQuery.Types.FieldFilter { Field = Field.ToFieldReference(), Op = (FieldOp) Op, Value = ValueSerializer.Serialize(context, Value) } };
+        }
+
+        private static UnaryOp GetUnaryOperator(object value, FieldOp op)
+        {
+            switch (value)
+            {
+                case null:
+                    return op switch
+                    {
+                        FieldOp.Equal => UnaryOp.IsNull,
+                        FieldOp.NotEqual => UnaryOp.IsNotNull,
+                        _ => throw new ArgumentException("Null values can only be used with the Equal/NotEqual operators", nameof(value))
+                    };
+                case double d when double.IsNaN(d):
+                case float f when float.IsNaN(f):
+                    return op switch
+                    {
+                        FieldOp.Equal => UnaryOp.IsNan,
+                        FieldOp.NotEqual => UnaryOp.IsNotNan,
+                        _ => throw new ArgumentException("Not-a-number values can only be used with the Equal/NotEqual operators", nameof(value))
+                    };
+                default:
+                    return UnaryOp.Unspecified;
+            }
+        }
+
+        public bool Equals(UnaryFilter other) => throw new NotImplementedException();
+    }
+
+    /// <summary>
+    /// 
+    /// </summary>
+    internal class CompositeFilter : Filter, IEquatable<CompositeFilter>
+    {
+        /// <summary>
+        /// 
+        /// </summary>
+        public IEnumerable<Filter> Filters { get; }

```

Comment:
I'd suggest making all the members of internal types internal as well, unless they're overriding public behavior. It makes it easier to review changes later. (Removing an internal member is fine; removing a public member *looks* like it's breaking, until you notice it's in an internal type.)